[RecordFunction] Store a c10::variant of name and schema rather then both.

robieta opened the pull request:
Summary: RecordFunction can be created either with a `c10::OperatorHandle` (from the dispatcher) or a string (everywhere else). We store a bunch of fields in RecordFunction to handle both paths, and it also complicates the logic since the input and output size can either be the size of `inputs_` and `outputs_` OR taken from the schema. (And that significantly complicates later changes) Because the dispatcher is the only place where we call the schema based method, we can just bind a reference and pass a `reference_wrapper`. (This is different from the other proposal that RecordFunction holds a schema pointer; in this case the caller just guarantees the schema for the lifetime of the guard.)

Test Plan: Ran the overhead benchmark. It helps quite a bit (0.37us -> 0.33us), presumably because there's just a lot less state (and assigns / dtors) in the guard.

Reviewed By: chaekit

Differential Revision: D35651041

albanD commented on torch/csrc/autograd/function.h:
```diff
@@ -151,24 +151,21 @@ struct TORCH_API Node : std::enable_shared_from_this<Node> {
-        } else {
-          guard.before(name(), sequence_nr());
-        }
+    auto step_callbacks = at::getStepCallbacks(at::RecordScope::BACKWARD_FUNCTION);
```
How expensive is this call? This is a very hot path.

Commit:
[RecordFunction] Store a c10::variant of name and schema rather then both. (#76017)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76017

RecordFunction can be created either with a `c10::OperatorHandle` (from the dispatcher) or a string (everywhere else). We store a bunch of fields in RecordFunction to handle both paths, and it also complicates the logic since the input and output size can either be the size of `inputs_` and `outputs_` OR taken from the schema. (And that significantly complicates later changes) Because the dispatcher is the only place where we call the schema based method, we can just bind a reference and pass a `reference_wrapper`. (This is different from the other proposal that RecordFunction holds a schema pointer; in this case the caller just guarantees the schema for the lifetime of the guard.)

Test Plan: Ran the overhead benchmark. It helps quite a bit (0.37us -> 0.33us), presumably because there's just a lot less state (and assigns / dtors) in the guard.

Reviewed By: chaekit

Differential Revision: D35651041

fbshipit-source-id: d6ba09f38b1d51c4c59872e6d35ebc6e85d758dc

Commit:
Merge branch 'master' into export-D35651041

robieta commented:
@pytorchbot merge this please
